File: packages/gavio-py/gavio/providers/_http.py
"""Tiny async JSON-over-HTTP helper built on stdlib (keeps core dependency-free)."""

from __future__ import annotations

import asyncio
import json
import urllib.error
import urllib.request

from ..exceptions import (
    ProviderUnavailableError,
    RateLimitError,
    ServerError,
)


async def post_json(
    url: str,
    payload: dict,
    headers: dict[str, str],
    timeout: float = 30.0,
) -> dict:
    """POST ``payload`` as JSON and return the parsed response.

    Runs the blocking request in a worker thread so callers stay async.
    Maps HTTP status families onto Gavio's transient error types so the
    retry/fallback policies can react.
    """

    def _do() -> dict:
        data = json.dumps(payload).encode("utf-8")
        req = urllib.request.Request(url, data=data, method="POST")
        req.add_header("Content-Type", "application/json")
        for key, value in headers.items():
            req.add_header(key, value)
        try:
            with urllib.request.urlopen(req, timeout=timeout) as resp:
                return json.loads(resp.read().decode("utf-8"))
        except urllib.error.HTTPError as error:
            body = error.read().decode("utf-8", errors="replace")
            if error.code == 429:
                raise RateLimitError(f"429 from provider: {body[:200]}") from error
            if error.code >= 500:
                raise ServerError(f"{error.code} from provider: {body[:200]}") from error
            raise ProviderUnavailableError(
                f"{error.code} from provider: {body[:200]}"
            ) from error
        except urllib.error.URLError as error:
            raise ProviderUnavailableError(f"network error: {error.reason}") from error

    return await asyncio.to_thread(_do)

File: packages/gavio-js/src/providers/http.ts
/** Tiny JSON-over-HTTP helper built on native fetch (keeps core dependency-free). */

import {
  ProviderUnavailableError,
  RateLimitError,
  ServerError,
} from '../errors.js'

/**
 * POST `payload` as JSON and return the parsed response.
 *
 * Maps HTTP status families onto Gavio's transient error types so the
 * retry/fallback policies can react.
 */
export async function postJson(
  url: string,
  payload: unknown,
  headers: Record<string, string>,
  timeoutSeconds = 30.0,
): Promise<Record<string, unknown>> {
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), timeoutSeconds * 1000)
  let resp: Response
  try {
    resp = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', ...headers },
      body: JSON.stringify(payload),
      signal: controller.signal,
    })
  } catch (error) {
    const reason = error instanceof Error ? error.message : String(error)
    throw new ProviderUnavailableError(`network error: ${reason}`)
  } finally {
    clearTimeout(timer)
  }

  if (!resp.ok) {
    const body = (await resp.text().catch(() => '')).slice(0, 200)
    if (resp.status === 429) {
      throw new RateLimitError(`429 from provider: ${body}`)
    }
    if (resp.status >= 500) {
      throw new ServerError(`${resp.status} from provider: ${body}`)
    }
    throw new ProviderUnavailableError(`${resp.status} from provider: ${body}`)
  }

  return (await resp.json()) as Record<string, unknown>
}
